Задание 1.
Выберите датасет, который имеет отношение к вашей области интересов или исследований. 
Датасет должен содержать неструктурированные данные, требующие разметки для решения конкретной задачи, например, анализа настроений или распознавания именованных сущностей.

In [1]:
# Импортируем библиотеки для скачивания с сайта Kaggle
from pathlib import Path
import opendatasets as od
# Импортируем библиотеки для работы с датасетом
import pandas as pd
import numpy as np
# Импортируем библиотеки для построения графиков
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Скачиваем дата сет с сайта Kaggle
dataset_path = Path('house-prices')
if not dataset_path.is_file():
    od.download('https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction')

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username:Your Kaggle Key:Dataset URL: https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction


100%|██████████| 8.56k/8.56k [00:00<00:00, 2.92MB/s]

Для начала выполним обработку DataSet

In [51]:
# Добавляем файл датасет в DataFrame pandas
df = pd.read_csv('C:\\Users\\Esdesu\\Desktop\\JreJre\\Сбор и разметка данных\\HomeWork\\Work#9\\heart-failure-prediction\\heart.csv')

In [52]:
# Анализируем данные
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [53]:
# Для дальнейшей работы приводим название колонок к нижнему регистру
df.columns = df.columns.str.lower()

In [54]:
# Обработка для столбца Age: age of the patient(Возраст пациента)
df['age'].info() # Определяем тип данных, количество значений
df['age'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: age
Non-Null Count  Dtype
--------------  -----
918 non-null    int64
dtypes: int64(1)
memory usage: 7.3 KB


array([40, 49, 37, 48, 54, 39, 45, 58, 42, 38, 43, 60, 36, 44, 53, 52, 51,
       56, 41, 32, 65, 35, 59, 50, 47, 31, 46, 57, 55, 63, 66, 34, 33, 61,
       29, 62, 28, 30, 74, 68, 72, 64, 69, 67, 73, 70, 77, 75, 76, 71],
      dtype=int64)

In [55]:
# Обработка для столбца Sex: sex of the patient(Пол пациента)
df['sex'].info() # Определяем тип данных, количество значений
df['sex'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: sex
Non-Null Count  Dtype 
--------------  ----- 
918 non-null    object
dtypes: object(1)
memory usage: 7.3+ KB


array(['M', 'F'], dtype=object)

In [56]:
# Обработка для столбца ChestPainType: chest pain type(Тип боли в груди)
df['chestpaintype'].info() # Определяем тип данных, количество значений
df['chestpaintype'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: chestpaintype
Non-Null Count  Dtype 
--------------  ----- 
918 non-null    object
dtypes: object(1)
memory usage: 7.3+ KB


array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [57]:
# Обработка для столбца RestingBP: resting blood pressure(Артериальное давление в состоянии покоя)
df['restingbp'].info() # Определяем тип данных, количество значений
df['restingbp'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: restingbp
Non-Null Count  Dtype
--------------  -----
918 non-null    int64
dtypes: int64(1)
memory usage: 7.3 KB


array([140, 160, 130, 138, 150, 120, 110, 136, 115, 100, 124, 113, 125,
       145, 112, 132, 118, 170, 142, 190, 135, 180, 108, 155, 128, 106,
        92, 200, 122,  98, 105, 133,  95,  80, 137, 185, 165, 126, 152,
       116,   0, 144, 154, 134, 104, 139, 131, 141, 178, 146, 158, 123,
       102,  96, 143, 172, 156, 114, 127, 101, 174,  94, 148, 117, 192,
       129, 164], dtype=int64)

In [58]:
# Смотрим значения переменной
df['restingbp'].describe().round(2)

count    918.00
mean     132.40
std       18.51
min        0.00
25%      120.00
50%      130.00
75%      140.00
max      200.00
Name: restingbp, dtype: float64

Обнаруженно одно 0, после анализа была принято решение привести значение к немного выше минимального.

In [59]:
# Сначала находим минимальное значение, исключая нули
min_value = df['restingbp'][df['restingbp'] > 0].min()
# Затем заменяем нулевые значения на значение чуть выше минимального
df['restingbp'] = df['restingbp'].replace(0, min_value * 1.01)

In [60]:
# Смотрим значения переменной
df['restingbp'].describe().round(2)

count    918.00
mean     132.48
std       18.07
min       80.00
25%      120.00
50%      130.00
75%      140.00
max      200.00
Name: restingbp, dtype: float64

In [61]:
# Обработка для столбца Cholesterol: serum cholesterol(Холестерин)
df['cholesterol'].info() # Определяем тип данных, количество значений
df['cholesterol'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: cholesterol
Non-Null Count  Dtype
--------------  -----
918 non-null    int64
dtypes: int64(1)
memory usage: 7.3 KB


array([289, 180, 283, 214, 195, 339, 237, 208, 207, 284, 211, 164, 204,
       234, 273, 196, 201, 248, 267, 223, 184, 288, 215, 209, 260, 468,
       188, 518, 167, 224, 172, 186, 254, 306, 250, 177, 227, 230, 294,
       264, 259, 175, 318, 216, 340, 233, 205, 245, 194, 270, 213, 365,
       342, 253, 277, 202, 297, 225, 246, 412, 265, 182, 218, 268, 163,
       529, 100, 206, 238, 139, 263, 291, 229, 307, 210, 329, 147,  85,
       269, 275, 179, 392, 466, 129, 241, 255, 276, 282, 338, 160, 156,
       272, 240, 393, 161, 228, 292, 388, 166, 247, 331, 341, 243, 279,
       198, 249, 168, 603, 159, 190, 185, 290, 212, 231, 222, 235, 320,
       187, 266, 287, 404, 312, 251, 328, 285, 280, 192, 193, 308, 219,
       257, 132, 226, 217, 303, 298, 256, 117, 295, 173, 315, 281, 309,
       200, 336, 355, 326, 171, 491, 271, 274, 394, 221, 126, 305, 220,
       242, 347, 344, 358, 169, 181,   0, 236, 203, 153, 316, 311, 252,
       458, 384, 258, 349, 142, 197, 113, 261, 310, 232, 110, 12

In [62]:
# Смотрим значения переменной
df['cholesterol'].describe().round(2)

count    918.00
mean     198.80
std      109.38
min        0.00
25%      173.25
50%      223.00
75%      267.00
max      603.00
Name: cholesterol, dtype: float64

Обнаруженно множество пустых данных 0, после анализа была принято решение заполнить данными немного выше среднего.

In [63]:
# Заполняем пустые значения средними значениями
df['cholesterol'] = df['cholesterol'].replace(0, df['cholesterol'].quantile(0.9))
# Проверяем
df['cholesterol'].unique()

array([289, 180, 283, 214, 195, 339, 237, 208, 207, 284, 211, 164, 204,
       234, 273, 196, 201, 248, 267, 223, 184, 288, 215, 209, 260, 468,
       188, 518, 167, 224, 172, 186, 254, 306, 250, 177, 227, 230, 294,
       264, 259, 175, 318, 216, 340, 233, 205, 245, 194, 270, 213, 365,
       342, 253, 277, 202, 297, 225, 246, 412, 265, 182, 218, 268, 163,
       529, 100, 206, 238, 139, 263, 291, 229, 307, 210, 329, 147,  85,
       269, 275, 179, 392, 466, 129, 241, 255, 276, 282, 338, 160, 156,
       272, 240, 393, 161, 228, 292, 388, 166, 247, 331, 341, 243, 279,
       198, 249, 168, 603, 159, 190, 185, 290, 212, 231, 222, 235, 320,
       187, 266, 287, 404, 312, 251, 328, 285, 280, 192, 193, 308, 219,
       257, 132, 226, 217, 303, 298, 256, 117, 295, 173, 315, 281, 309,
       200, 336, 355, 326, 171, 491, 271, 274, 394, 221, 126, 305, 220,
       242, 347, 344, 358, 169, 181, 236, 203, 153, 316, 311, 252, 458,
       384, 258, 349, 142, 197, 113, 261, 310, 232, 110, 123, 17

In [64]:
# Смотрим значения переменной
df['cholesterol'].describe().round(2)

count    918.00
mean     255.95
std       58.29
min       85.00
25%      214.00
50%      254.00
75%      305.00
max      603.00
Name: cholesterol, dtype: float64

In [65]:
# Обработка для столбца FastingBS: fasting blood sugar(Уровень сахара в крови натощак(Диабет))
df['fastingbs'].info() # Определяем тип данных, количество значений
df['fastingbs'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: fastingbs
Non-Null Count  Dtype
--------------  -----
918 non-null    int64
dtypes: int64(1)
memory usage: 7.3 KB


array([0, 1], dtype=int64)

In [66]:
# Обработка для столбца RestingECG: resting electrocardiogram results(Результаты электрокардиограммы покоя)
df['restingecg'].info() # Определяем тип данных, количество значений
df['restingecg'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: restingecg
Non-Null Count  Dtype 
--------------  ----- 
918 non-null    object
dtypes: object(1)
memory usage: 7.3+ KB


array(['Normal', 'ST', 'LVH'], dtype=object)

In [67]:
# Обработка для столбца MaxHR: maximum heart rate achieved(Максимальная частота пульса)
df['maxhr'].info() # Определяем тип данных, количество значений
df['maxhr'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: maxhr
Non-Null Count  Dtype
--------------  -----
918 non-null    int64
dtypes: int64(1)
memory usage: 7.3 KB


array([172, 156,  98, 108, 122, 170, 142, 130, 120,  99, 145, 140, 137,
       150, 166, 165, 125, 160, 164, 138, 178, 112, 118, 127, 114, 154,
       155,  87, 148, 100, 168, 184, 121, 153, 134,  96, 174, 175, 144,
        82, 135, 115, 128, 116,  94, 110,  92, 180, 152, 124, 106, 185,
       139, 190, 146, 158, 132, 176, 119, 188, 162, 105,  90, 136, 167,
       129, 102, 143, 103,  91, 126,  93, 131, 149, 123, 182, 141,  77,
       109, 133, 179, 113, 104,  95,  72,  97, 117,  86,  63, 157,  83,
        60,  70, 163,  67,  78,  84, 111,  80, 107, 161,  69,  88,  73,
       159, 151, 181, 186, 177, 173, 169, 171, 147,  71, 192, 195, 194,
       187, 202], dtype=int64)

In [68]:
# Обработка для столбца ExerciseAngina: exercise-induced angina(Стенокардия, вызванная физической нагрузкой)
df['exerciseangina'].info() # Определяем тип данных, количество значений
df['exerciseangina'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: exerciseangina
Non-Null Count  Dtype 
--------------  ----- 
918 non-null    object
dtypes: object(1)
memory usage: 7.3+ KB


array(['N', 'Y'], dtype=object)

In [69]:
# Обработка для столбца Oldpeak
df['oldpeak'].info() # Определяем тип данных, количество значений
df['oldpeak'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: oldpeak
Non-Null Count  Dtype  
--------------  -----  
918 non-null    float64
dtypes: float64(1)
memory usage: 7.3 KB


array([ 0. ,  1. ,  1.5,  2. ,  3. ,  4. ,  0.5,  2.5,  5. ,  0.8,  0.7,
        1.4,  2.1,  0.4,  0.2,  1.7,  2.2,  0.1,  1.6,  1.3,  0.3,  1.8,
        2.6, -0.9,  2.8, -2.6, -1.5, -0.1,  0.9,  1.1,  2.4, -1. , -1.1,
       -0.7, -0.8,  3.7,  1.2, -0.5, -2. ,  1.9,  3.5,  0.6,  3.1,  2.3,
        3.4,  3.6,  4.2,  3.2,  5.6,  3.8,  2.9,  6.2,  4.4])

In [70]:
# Обработка для столбца ST_Slope: the slope of the peak exercise ST segment(Наклон пикового сегмента ST при нагрузке)
df['st_slope'].info() # Определяем тип данных, количество значений
df['st_slope'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: st_slope
Non-Null Count  Dtype 
--------------  ----- 
918 non-null    object
dtypes: object(1)
memory usage: 7.3+ KB


array(['Up', 'Flat', 'Down'], dtype=object)

In [71]:
# Обработка для столбца HeartDisease(Болезнь сердца)
df['heartdisease'].info() # Определяем тип данных, количество значений
df['heartdisease'].unique() # Поиск уникальных значений

<class 'pandas.core.series.Series'>
RangeIndex: 918 entries, 0 to 917
Series name: heartdisease
Non-Null Count  Dtype
--------------  -----
918 non-null    int64
dtypes: int64(1)
memory usage: 7.3 KB


array([0, 1], dtype=int64)

In [72]:
# Анализируем типы данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             918 non-null    int64  
 1   sex             918 non-null    object 
 2   chestpaintype   918 non-null    object 
 3   restingbp       918 non-null    float64
 4   cholesterol     918 non-null    int64  
 5   fastingbs       918 non-null    int64  
 6   restingecg      918 non-null    object 
 7   maxhr           918 non-null    int64  
 8   exerciseangina  918 non-null    object 
 9   oldpeak         918 non-null    float64
 10  st_slope        918 non-null    object 
 11  heartdisease    918 non-null    int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 86.2+ KB


Dataset Description

1. Age: age of the patient [years]
2. Sex: sex of the patient [M: Male, F: Female]
3. ChestPainType: chest pain type [TA: Typical Angina, ATA: Atypical Angina, NAP: Non-Anginal Pain, ASY: Asymptomatic]
4. RestingBP: resting blood pressure [mm Hg]
5. Cholesterol: serum cholesterol [mm/dl]
6. FastingBS: fasting blood sugar [1: if FastingBS > 120 mg/dl, 0: otherwise]
7. RestingECG: resting electrocardiogram results [Normal: Normal, ST: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV), LVH: showing probable or definite left ventricular hypertrophy by Estes' criteria]
8. MaxHR: maximum heart rate achieved [Numeric value between 60 and 202]
9. ExerciseAngina: exercise-induced angina [Y: Yes, N: No]
10. Oldpeak: oldpeak = ST [Numeric value measured in depression]
11. ST_Slope: the slope of the peak exercise ST segment [Up: upsloping, Flat: flat, Down: downsloping]
12. HeartDisease: output class [1: heart disease, 0: Normal]

Описание набора данных

1. Возраст: возраст пациента [лет]
2. Пол: пол пациента [М: Мужчина, Ж: Женщина]
3. ChestPainType: тип боли в груди [TA: типичная стенокардия, ATA: атипичная стенокардия, NAP: неангинальная боль, ASY: бессимптомная]
4. RestingBP: артериальное давление в состоянии покоя [мм рт. ст.]
5. Холестерин: холестерин сыворотки [мм/дл]
6. FastingBS: уровень сахара в крови натощак [1: если FastingBS > 120 мг/дл, 0: в противном случае]
7. ЭКГ покоя: результаты электрокардиограммы покоя [Нормальный: нормальный, ST: наличие аномалий ST-T (инверсия зубца T и/или подъем или депрессия ST > 0,05 мВ), ГЛЖ: вероятная или определенная гипертрофия левого желудочка по критериям Эстеса ]
8. MaxHR: достигнутая максимальная частота пульса [числовое значение от 60 до 202]
9. Стенокардия при физической нагрузке: стенокардия, вызванная физической нагрузкой [Д: Да, Н: Нет]
10. Oldpeak: oldpeak = ST [числовое значение, измеренное в депрессии]
11. ST_Slope: наклон пикового сегмента ST при нагрузке [Вверх: наклон вверх, Плоский: плоский, Вниз: наклон вниз]
12. HeartDisease: выходной класс [1: заболевание сердца, 0: нормальное]

Данный DataSet представляет собой сбор данных сердечно-сосудистых заболеваний. Разработываемая модель должна анализировать данные и сама определять риск заболевания, для пациента на основе собраных данных исследований.

Задание 2.
Выполните разметку на основе правил (rule-based labeling) на подмножестве выбранного датасета. 
Разработайте и реализуйте набор правил или условий, которые позволят автоматически присваивать метки данным на основе определенных шаблонов или критериев.

In [73]:
# Для удобства разделим DataSet на два для разметки автоматически 95% и вручную 5%
# Импорт модуля библиотеки для разделения df
from sklearn.model_selection import train_test_split

df_auto, df_manual = train_test_split(df, test_size=0.05, random_state=42)

In [83]:
def risk_disease(row, age, chestpaintype, fastingbs, st_slope, heartdisease):
    if (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'ASY' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'BA' # Risk Disease Below Average
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'ASY' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'H' # Risk Disease High
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'ASY' and row[st_slope] == 'Down' and row[heartdisease] != 1:
        return 'VH' # Risk Disease Very High
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'NAP' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'M' # Risk Disease Minimal
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'NAP' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'A' # Risk Disease Average
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'NAP' and row[fastingbs] != 1 and row[st_slope] == 'Down' and row[heartdisease] != 1:
        return 'M' # Risk Disease Minimal
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'ATA' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'VM' # Risk Disease Very Minimal
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'ATA' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'BA' # Risk Disease Below Average
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'ATA' and row[st_slope] == 'Down' and row[heartdisease] != 1:
        return 'VH' # Risk Disease Very High
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'TA' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'VM' # Risk Disease Very Minimal
    elif (row[age] >= 28 and row[age] <= 50) and row[chestpaintype] == 'TA' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'VH' # Risk Disease Very High 
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'ASY' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'A' # Risk Disease Average
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'ASY' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'H' # Risk Disease Hing
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'ASY' and row[st_slope] == 'Down' and row[heartdisease] != 1:
        return 'VH' # Risk Disease Very High
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'NAP' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'VM' # Risk Disease Very Minimal
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'NAP' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'M' # Risk Disease Minimal
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'NAP' and row[st_slope] == 'Down' and row[heartdisease] != 1:
        return 'H' # Risk Disease Hing
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'ATA' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'VM' # Risk Disease Very Minimal
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'ATA' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'AA' # Risk Above Average
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'ATA' and row[st_slope] == 'Down' and row[heartdisease] != 1:
        return 'VM' # Risk Disease Very Minimal
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'TA' and row[st_slope] == 'Up' and row[heartdisease] != 1:
        return 'VM' # Risk Disease Very Minimal
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'TA' and row[st_slope] == 'Flat' and row[heartdisease] != 1:
        return 'A' # Risk Disease Average
    elif (row[age] >= 51 and row[age] <= 77) and row[chestpaintype] == 'TA' and row[st_slope] == 'Down' and row[heartdisease] != 1:
        return 'VM' # Risk Disease Very Minimal
    else:
        return 'HD' # Heart Disease

df_auto['risk_disease'] = df_auto.apply(lambda row: risk_disease(row, 'age', 'chestpaintype', 'fastingbs', 'st_slope', 'heartdisease'), axis = 1)

In [87]:
df_auto.head(33)

age sex chestpaintype  restingbp  cholesterol  fastingbs restingecg  \
250   44   M           ASY      135.0          491          0     Normal   
752   56   M           ASY      125.0          249          1        LVH   
350   53   M           ASY      120.0          305          1     Normal   
758   51   M            TA      125.0          213          0        LVH   
759   54   M           ATA      192.0          283          0        LVH   
107   34   M           ATA      150.0          214          0         ST   
445   55   M           NAP      136.0          228          0         ST   
141   50   M           ASY      140.0          341          0         ST   
650   48   M           ASY      130.0          256          1        LVH   
544   61   F           ATA      140.0          298          1     Normal   
110   59   F           ATA      130.0          188          0     Normal   
593   64   M           ASY      130.0          258          1        LVH   
519   63   M           ASY       96.0          305          0         ST   
907   44   M           ASY      120.0          169          0     Normal   
675   57   M           NAP      150.0          126          1     Normal   
280   60   M           NAP      120.0          246          0        LVH   
136   43   F           ATA      120.0          215          0         ST   
422   65   M           ASY      150.0          236          1         ST   
208   28   M           ATA      130.0          132          0        LVH   
442   51   M           ASY      128.0          305          1         ST   
86    65   M           ASY      170.0          263          1     Normal   
44    43   M           ASY      120.0          175          0     Normal   
531   64   M           ASY      143.0          306          1         ST   
913   45   M            TA      110.0          264          0     Normal   
634   40   M            TA      140.0          199          0     Normal   
290   48   M           NAP      110.0          211          0     Normal   
338   63   M           ASY      140.0          305          1        LVH   
357   53   M           ATA      120.0          305          0     Normal   
292   53   M           ASY      130.0          182          0     Normal   
227   38   M           ASY       92.0          117          0     Normal   
591   58   M           ASY      100.0          213          0         ST   
425   60   M           ATA      160.0          267          1         ST   
789   34   M            TA      118.0          182          0        LVH   

     maxhr exerciseangina  oldpeak st_slope  heartdisease risk_disease  
250    135              N      0.0     Flat             1           HD  
752    144              Y      1.2     Flat             1           HD  
350    120              N      0.0     Flat             1           HD  
758    125              Y      1.4       Up             0           VM  
759    195              N      0.0       Up             1           HD  
107    168              N      0.0       Up             0           VM  
445    124              Y      1.6     Flat             1           HD  
141    125              Y      2.5     Flat             1           HD  
650    150              Y      0.0       Up             1           HD  
544    120              Y      0.0       Up             0           VM  
110    124              N      1.0     Flat             0           AA  
593    130              N      0.0     Flat             1           HD  
519    121              Y      1.0       Up             1           HD  
907    144              Y      2.8     Down             1           HD  
675    173              N      0.2       Up             0           VM  
280    135              N      0.0       Up             0           VM  
136    175              N      0.0       Up             0           VM  
422    105              Y      0.0     Flat             1           HD  
208    185              N      0.

In [88]:
# Проверяем тип и заполнение
df_auto.info()

<class 'pandas.core.frame.DataFrame'>
Index: 872 entries, 250 to 102
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             872 non-null    int64  
 1   sex             872 non-null    object 
 2   chestpaintype   872 non-null    object 
 3   restingbp       872 non-null    float64
 4   cholesterol     872 non-null    int64  
 5   fastingbs       872 non-null    int64  
 6   restingecg      872 non-null    object 
 7   maxhr           872 non-null    int64  
 8   exerciseangina  872 non-null    object 
 9   oldpeak         872 non-null    float64
 10  st_slope        872 non-null    object 
 11  heartdisease    872 non-null    int64  
 12  risk_disease    872 non-null    object 
dtypes: float64(2), int64(5), object(6)
memory usage: 95.4+ KB



Задача 3.
Выполните разметку вручную отдельного подмножества выбранного датасета с помощью выбранного вами инструмента разметки.

Задача 4.
Объедините данные, размеченные вручную, с данными, размеченными на основе правил. 
Объедините два подмножества размеченных данных в один набор данных, сохранив при этом соответствующую структуру и целостность.

Задача 5.
Обучите модель машинного обучения, используя объединенный набор размеченных данных. 
Разделите датасет на обучающий и тестовый наборы и используйте обучающий набор для обучения модели.

Задача 6.
Оценить эффективность обученной модели на тестовом датасете. Используйте подходящие метрики оценки. 
Интерпретируйте результаты и проанализируйте эффективность модели в решении задачи разметки.